src/linux_mcp_server/tools/system_info.py

```diff
@@ -350,156 +511,272 @@ async def get_memory_information(
+
+    return device
+
+
+def _parse_usb_device_attrs(device_name: str, output: str) -> USBDevice:
+    """Parse USB device attributes from raw output."""
+    parts = device_name.split("-")
+    bus = parts[0]
+    device = USBDevice(bus=bus, device=device_name)
+
+    for line in output.strip().split("\n"):
+        if ":" in line:
+            key, value = line.split(":", 1)
+            value = value.strip()
+            if value:
+                match key:
+                    case "idVendor":
+                        device.vendor_id = value
+                    case "idProduct":
+                        device.product_id = value
+                    case "product":
+                        device.description = value
+                    case "manufacturer":
+                        device.manufacturer = value
+
+    return device
+
+
+async def _parse_device_information(raw_outputs: dict[CommandKey, RawCommandOutput]) -> ParsedData:
+    """
+    Parse phase: Convert raw device outputs into structured device data.
+
+    Returns a dictionary with pci_devices and usb_devices lists.
+    """
+    pci_devices = []
+    usb_devices = []
+
+    # Parse PCI and USB devices
+    for command_key, output in raw_outputs.items():
+        if command_key[0] == "pci_attrs" and len(command_key) == 2:
+            address = command_key[1]
+            pci_devices.append(_parse_pci_device_attrs(address, output))
+        elif command_key[0] == "usb_attrs" and len(command_key) == 2:
+            device_name = command_key[1]
+            usb_devices.append(_parse_usb_device_attrs(device_name, output))
+
+    return {
+        "pci_devices": pci_devices,
+        "usb_devices": usb_devices,
+    }
+
+
+async def _filter_device_information(
+    parsed_data: ParsedData, device_types: list[str] | None = None, limit: int | None = None
+) -> ParsedData:
+    """
+    Filter phase: Apply device type and limit filters to device data.
+
+    Args:
+        parsed_data: The parsed device data containing pci_devices and usb_devices
+        device_types: List of device types to include ('pci', 'usb'), or None for all
+        limit: Maximum number of devices to return per type, or None for all
+
+    Returns:
+        Filtered device data
+    """
+    filtered = {}
+
+    # Apply device type filter
+    should_include_pci = device_types is None or "pci" in device_types
+    should_include_usb = device_types is None or "usb" in device_types
+
+    if should_include_pci and "pci_devices" in parsed_data:
+        pci_devices = parsed_data["pci_devices"]
+        if limit is not None:
+            pci_devices = pci_devices[:limit]
+        filtered["pci_devices"] = pci_devices
+
+    if should_include_usb and "usb_devices" in parsed_data:
+        usb_devices = parsed_data["usb_devices"]
+        if limit is not None:
+            usb_devices = usb_devices[:limit]
+        filtered["usb_devices"] = usb_devices
+
+    return filtered
```

Comment:
Again, it would be better to return an object unless that's really not feasible.